python/ray/tune/impl/tuner_internal.py

```diff
@@ -192,8 +195,83 @@ def _maybe_warn_resource_contention(self):
                 stacklevel=4,
             )
 
+    def _validate_overwrite_trainable(
+        self,
+        original_trainable: TrainableTypeOrTrainer,
+        overwrite_trainable: Optional[TrainableTypeOrTrainer],
+    ):
+        """Determines whether the new `overwrite_trainable` is compatible
+        with the restored experiment with some basic sanity checks
+        (ensuring same type and name as the original trainable).
+        """
+
+        # Check if the trainable was wrapped with `tune.with_parameters`,
+        # Set the Tuner to fail on fit if the trainable is not re-specified.
+        trainable_wrapped_params = getattr(
+            original_trainable, "_attached_param_names", None
+        )
+        if trainable_wrapped_params and not overwrite_trainable:
+            self._missing_params_error_message = (
+                "The original trainable cannot be used to resume training, since "
+                "`tune.with_parameters` attached references to objects "
+                "in the Ray object store that may not exist anymore. "
+                "You must re-supply the trainable with the same parameters "
+                f"{trainable_wrapped_params} attached:\n\n"
+                "from ray import tune\n\n"
+                "# Reconstruct the trainable with the same parameters\n"
+                "trainable_with_params = tune.with_parameters(trainable, ...)\n"
+                "tuner = tune.Tuner.restore(\n"
+                "    ..., overwrite_trainable=trainable_with_params\n"
+                ")\n\nSee https://docs.ray.io/en/master/tune/api_docs/trainable.html"
+                "#tune-with-parameters for more details."
+            )
+        if not overwrite_trainable:
+            return
+
+        error_message = (
+            "Usage of `overwrite_trainable` is limited to re-specifying the "
+            "same trainable that was passed to `Tuner`, in the case "
+            "that the trainable is not serializable (e.g. it holds object references)."
+        )
+
+        if type(original_trainable) != type(overwrite_trainable):
+            raise ValueError(
+                f"{error_message}\n"
+                f"Got new trainable of type {type(overwrite_trainable)} "
+                f"but expected {type(original_trainable)}."
+            )
+
+        from ray.train.trainer import BaseTrainer
+
+        if isinstance(overwrite_trainable, BaseTrainer):
+            if overwrite_trainable.run_config != original_trainable.run_config:
+                logger.warning(
```

Comment:
this one should be warnings.warn since the user can change their code to address this?